Simple mdns implementation using the Espressif code

pjsg opened the pull request:
This PR is a thin layer over the espressif provided mdns implementation. It only allows a single service to be advertised and it doesn't allow the hostname and the instance name to be broken apart. However, it works, is simple, and reasonably compact.

marcelstoer commented on app/include/user_modules.h:
```diff
@@ -27,6 +27,7 @@
 #define LUA_USE_MODULES_HTTP
 //#define LUA_USE_MODULES_HX711
 #define LUA_USE_MODULES_I2C
+#define LUA_USE_MODULES_MDNS
```
I'd rather have this disabled by default even though #872 is still waiting for a resolution.

Alkorin commented on app/modules/mdns.c:
```diff
@@ -0,0 +1,130 @@
+  char *p = NULL;
+
+  if (phase == 1) {
+    result = (struct wrapper *) c_zalloc(sizeof(struct wrapper) + *sizep);
```
Never freed ?

Alkorin commented:
It tooks me more than 10 minutes to understand what you tried to do with your phase 0 / phase 1.

You could split this func in two parts with more comments, it'll add readability to your code.
setptr is not obvious too... You use only 3 times, it's maybe better to write it like this

```
...
tmpBuffer = &result->data;
...
// Get server name
char * server_name = luaL_checkstring(L, 2)
// Store it
c_strcpy(tmpBuffer, server_name);
result->mdns_info.server_name = tmpBuffer;
tmpBuffer += c_strlen(server_name) + 1;
```
than
```
setptr(&result->mdns_info.server_name, &p, luaL_checkstring(L, 2));
```

It's easier to read.

pjsg commented:
I cleaned up the code a bit and removed the setptr function. 

Also, I disabled the module by default. Turns out that the lwip code adds a fair amount of RAM consumption (maybe 700 bytes) if the mdns code is included.

makefu commented:
i'd love to see this in mainline :+1:

saeugetier commented:
I would like to see it in the mainline, too. It's a nice feature. In an DHCP environment you don't know each IP of each nodemcu module.

pastukhov commented:
:+1:

pastukhov commented:
Any progress?

pastukhov commented:
Hi! @pjsg.
I have cherrypicked tis pr, but can't get this working.
I have add ```lua mdns.register('socket', "http", 80, { hardware='NodeMCU'})``` to my http server startup script and not able to find socket.local with avahi-discover or dig @224.0.0.251 -p 5353 socket.local

pjsg commented:
I just built the branch again (against the current dev branch) and ran your mdns command line.

I then ran avahi-browse on my linux box:

avahi-browse --resolve _http._tcp
+ eth0 IPv4 socket                                        Web Site             local
+ eth0 IPv4 Living Room                                   Web Site             local
= eth0 IPv4 socket                                        Web Site             local
   hostname = [socket.local]
   address = [192.168.1.234]
   port = [80]
   txt = ["hardware=NodeMCU" "vendor = Espressif"]

You can see that 'socket' appears. Initially it didn't work until I modified my iptables configuration on my linux box to permit inbound UDP/5353 traffic. Note that the response doesn't come from the same address as the original packet destination, so the default return path doesn't work. Thus you need an explicit allow rule.

Also note that the service type is being prefixed by an underscore, so it doesn't matter if you include it in the registration or not.

Also, I think that you need to do the registration *after* you are connected to the network.

FrankX0 commented:
I am using mDNS with the latest dev build, but cannot get it stable to work.
The first few minutes the functionality works as expected, but after 5 to 10 minutes the service seems to stop. Through a web-browser I can no longer connect using the .local domain. Using dns-sd in Windows 10 also does not result in finding the IP address.
Any suggestions?
